djangocms_picture/models.py

```diff
@@ -1,83 +1,284 @@
+    )
+    # upscale and crop work together
+    # throws validation error if other cropping options are selected
+    use_crop = models.BooleanField(
+        verbose_name=_('Crop image'),
+        blank=True,
+        default=False,
+        help_text=_('Crops the image according to the given thumbnail settings in the template.'),
+    )
+    use_upscale = models.BooleanField(
+        verbose_name=_('Upscale image'),
+        blank=True,
+        default=False,
+        help_text=_('Upscales the image to the size of the thumbnail settings in the template.')
+    )
+    # overrides all other options
+    # throws validation error if other cropping options are selected
+    thumbnail_options = models.ForeignKey(
+        ThumbnailOption,
+        verbose_name=_('Thumbnail options'),
+        blank=True,
+        null=True,
+        help_text=_('Overrides width, height, crop and upscale with the provided preset.'),
+    )
+
+    # Add an app namespace to related_name to avoid field name clashes
+    # with any other plugins that have a field with the same name as the
+    # lowercase of the class name of this model.
+    # https://github.com/divio/django-cms/issues/5030
+    cmsplugin_ptr = models.OneToOneField(
+        CMSPlugin,
+        related_name='%(app_label)s_%(class)s',
+        parent_link=True,
+    )
 
     def __str__(self):
-        if self.alt:
-            return self.alt[:40]
-        elif self.image:
-            # added if, because it raised attribute error when file wasn't
-            # defined.
-            try:
-                return u"%s" % os.path.basename(self.image.name)
-            except AttributeError:
-                pass
-        return u"<empty>"
+        if self.picture and self.picture.label:
+            return self.picture.label
+        return str(self.pk)
+
+    def get_short_description(self):
+        if self.picture and self.picture.label:
+            return self.picture.label
+        if self.link_url:
+            return self.link_url
+        return ugettext('<file is missing>')
+
+    def copy_relations(self, oldinstance):
+        # Because we have a ForeignKey, it's required to copy over
+        # the reference from the instance to the new plugin.
+        self.picture = oldinstance.picture
+
+    def get_size(self, width=None, height=None):
+        crop = self.use_crop
+        upscale = self.use_upscale
+        # use field thumbnail settings
+        if self.thumbnail_options:
+            width = self.thumbnail_options.width
+            height = self.thumbnail_options.height
+            crop = self.thumbnail_options.crop
+            upscale = self.thumbnail_options.upscale
+        elif not self.use_automatic_scaling:
+            width = self.width
+            height = self.height
+
+        # calculate height when not given according to the
+        # golden ratio or fallback to the picture size
+        if not height and width:
```

Comment:
this can be rewritten as:

```python
if not height and width:
    height = int(width / PICTURE_RATIO)
elif not width and height:
    width = int(height * PICTURE_RATIO)
elif not width and not height:
    width = self.picture.width
    height = self.picture.height
```